### PR TITLE
Refactor wait_agent using ScEvent

### PR DESCRIPTION
**Pull request by FalseR20**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -735,9 +735,10 @@ def call_agent(
 ```
 
 Agent wait function: Waits for creation of edge to reaction node for some seconds.
+Default reaction_node is `question_finished`.
 
 ```python
-def wait_agent(seconds: float, question_node: ScAddr, reaction_node: ScAddr) -> None: ...
+def wait_agent(seconds: float, question_node: ScAddr, reaction_node: ScAddr = None) -> None: ...
```
Comment (AlexandrZagorskiy) [inline README.md:741]: In the description you say that default reaction_node is `question_finished`, but in the code you add None as the value of node by default

File: src/sc_kpm/utils/action_utils.py [hunk 2]
```diff
@@ -132,11 +135,23 @@ def call_action(action_node: ScAddr, initiation: Idtf = QuestionStatus.QUESTION_
     create_edge(sc_types.EDGE_ACCESS_CONST_POS_PERM, initiation_node, action_node)
 
 
-# TODO rewrite to event
-def wait_agent(seconds: float, question_node: ScAddr, reaction_node: ScAddr) -> None:
+def wait_agent(seconds: float, question_node: ScAddr, reaction_node: ScAddr = None) -> None:
+    reaction_node = reaction_node or ScKeynodes[QuestionStatus.QUESTION_FINISHED]
```
[the remark this answers is not in the document]
Comment (FalseR20) [reply src/sc_kpm/utils/action_utils.py:139]: I tried not to add [mutable default parameter](https://www.geeksforgeeks.org/default-arguments-in-python/#:~:text=Using%20mutable%20objects%20as%20default%20argument%20values%20in%20python).
But here this parameter doesn't change, so it won't broke logic

[the remark this answers is not in the document]
Comment (FalseR20) [reply src/sc_kpm/utils/action_utils.py:139]: Yeah, cannot without it
```
../venv/lib/python3.8/site-packages/sc_client/session.py:153: in send_message
    raise BrokenPipeError
E   BrokenPipeError
```

File: README.md (hunk 1 above)
Comment (FalseR20) [reply README.md:741]: Bottom

File: src/sc_kpm/utils/action_utils.py [hunk 3]
```diff
@@ -132,11 +135,23 @@ def call_action(action_node: ScAddr, initiation: Idtf = QuestionStatus.QUESTION_
...
+        nonlocal is_finished
+        is_finished = True
+        return ScResult.OK
+
+    event_params = ScEventParams(question_node, ScEventType.ADD_INGOING_EDGE, event_callback)
+    sc_event = events_create(event_params)[0]
     finish = datetime.now() + timedelta(seconds=seconds)
-    while not check_edge(sc_types.EDGE_ACCESS_VAR_POS_PERM, reaction_node, question_node) and datetime.now() < finish:
+    while not is_finished and datetime.now() < finish:
         time.sleep(0.1)
```
[the remark this answers is not in the document]
Comment (FalseR20) [reply src/sc_kpm/utils/action_utils.py]: Thank you, good solution

File: src/sc_kpm/utils/action_utils.py [hunk 4]
```diff
@@ -132,11 +134,20 @@ def call_action(action_node: ScAddr, initiation: Idtf = QuestionStatus.QUESTION_
...
+
+    def event_callback(_: ScAddr, __: ScAddr, trg: ScAddr) -> ScResult:
+        if trg != reaction_node:
+            return ScResult.SKIP
+        finish_event.set()
+        return ScResult.OK
+
+    event_params = ScEventParams(question_node, ScEventType.ADD_INGOING_EDGE, event_callback)
+    sc_event = events_create(event_params)[0]
+    finish_event.wait(seconds)  # TODO: return status in 0.2.0
```
Comment (FalseR20) [inline src/sc_kpm/utils/action_utils.py:149]: I don't want add logic of returning status in release 0.1.1

File: src/sc_kpm/utils/action_utils.py [hunk 5]
```diff
@@ -132,11 +134,20 @@ def call_action(action_node: ScAddr, initiation: Idtf = QuestionStatus.QUESTION_
...
+    events_destroy(sc_event)
```
Comment (AlexandrZagorskiy) [inline src/sc_kpm/utils/action_utils.py:150]: Could you add some tests after integrating such logic? I didn't found any tests for **wait_agent()** func. I'm not sure that  your version of waiting will be interrupted when agent will be finished

Comment (NikitaZotov) [reply src/sc_kpm/utils/action_utils.py:150]: Agree with previous reviewer

Comment (FalseR20) [reply src/sc_kpm/utils/action_utils.py:150]: `tests/test_utils/test_action_utils.py` - There are usages when it finishes and doeesn't.
Coverage: function covered on 100% (green bar on the left)
image

Comment (AlexandrZagorskiy) [reply src/sc_kpm/utils/action_utils.py:150]: image
I expect a separate testcase for a wait_agent(). It's hard to looks on coverage of this function when you review code via GitHub, this is not obvious. @NikitaZotov what do you think about it?

Comment (NikitaZotov) [reply src/sc_kpm/utils/action_utils.py:152]: Agree with you

Comment (FalseR20) [reply src/sc_kpm/utils/action_utils.py:152]: Test added

**Review by AlexandrZagorskiy — APPROVED** [review]

**Pull request merged**